Configurable default_ttl

coding-bunny opened the pull request:
Okay,

as requested in the issue I opened, I tried making a pull-request to make the default_ttl configurable or fallback to one year when not provided.
I hard a bit of trouble understanding how it works, but I added specs and they passed, without breaking the previous specs.

tubbo commented on lib/rack/cache/redis_entitystore.rb:
```diff
@@ -8,21 +7,26 @@ class RedisBase < self
 
-        def self.resolve(uri)
-          new ::Redis::Store::Factory.resolve(uri.to_s)
+        def self.resolve(uri, options = {})
```
Is this actually how the Rack::Cache API works?

tubbo commented on lib/rack/cache/redis_entitystore.rb:
```diff
@@ -8,21 +7,26 @@ class RedisBase < self
         # The underlying ::Redis instance used to communicate with the Redis daemon.
         attr_reader :cache
 
+        class << self
```
Not sure if this is the best way to configure default_ttl. I would much rather continue to use `Rack::Cache::DEFAULT_TTL` if a TTL is not provided in the initial configuration, then set `@default_ttl` on the actual instance each time based on whether an override was given.

tubbo commented on lib/rack/cache/redis_entitystore.rb:
```diff
@@ -8,21 +7,26 @@ class RedisBase < self
       class Redis < RedisBase
         def initialize(server, options = {})
           @cache = ::Redis::Store::Factory.create(server)
+          self.class.default_ttl = options[:default_ttl] || 86_400 * 365 # 1 year
```
As stated above, this looks like it really wants to be an instance variable. Also, you should still default this to `DEFAULT_TTL` if nothing has been given...

```ruby
@default_ttl = options[:default_ttl] || ::Redis::Rack::Cache::DEFAULT_TTL
```

tubbo commented on lib/rack/cache/redis_entitystore.rb:
```diff
@@ -37,7 +41,7 @@ def write(body, ttl=0)
           key, size = slurp(body){|part| buf.write(part) }
 
-          ttl = ::Redis::Rack::Cache::DEFAULT_TTL if ttl.zero?
+          ttl = self.class.default_ttl if ttl.zero?
```
This could just be set to `@default_ttl if ttl.zero?`

tubbo commented on test/rack/cache/entitystore/redis_test.rb:
```diff
@@ -11,6 +11,21 @@ def sha_like?
     @store = ::Rack::Cache::EntityStore::Redis.new :host => 'localhost'
   end
 
+  it 'has a default_tll of 1 year' do
```
These tests just verify that you can indeed set the TTL, they do not verify whether the TTL was used in calls to redis-store.

coding-bunny commented:
So changed the following:

* Switched to using an instance variable
* Undid the changes to `self.resolve` to respect the Rack API
* Added a snippet to the readme, even if we encourage people to use the redis-rails gem.

tubbo commented on README.md:
```diff
@@ -11,7 +11,22 @@ gem 'redis-rack-cache'
+module MyApplication
+  class Application < Rails::Application
+    config.action_dispatch.rack_cache = {
+      metastore: ::Rack::Cache::EntityStore::Redis.new('redis://localhost:6379/0/metastore', default_ttl: 10.days.to_i),
```
rack-cache is set up a little differently from this: https://github.com/rtomayko/rack-cache/blob/master/lib/rack/cache/options.rb#L38-L79

notice how `default_ttl` is on the same level as metastore and entitystore? I think this is what the hash of options needs to look like:

```ruby
config.action_dispatch.rack_cache = {
  metastore: 'redis://localhost:6379/0/rack-metastore',
  entitystore: 'redis://localhost:6379/0/rack-entitystore',
  default_ttl: 15.minutes.from_now
}
```

coding-bunny replied:
Neither approaches are actually working. I've tested it locally by pointing the Gemfile to my git repository, and whether I use my notation or the one recommended, the TTL keeps falling back to one year. Meaning the information is not passed down.

coding-bunny commented:
putting this on hold for now till I have a bit more time to dive into this.

tubbo commented:
Hey @coding-bunny there was some refactoring work I wanted to do with regards to `RedisBase`, and I believe there was a problem with the ancestor tree of those objects, hence the default_ttl issues. I'm going to close this PR and open a new one that includes both your changes and mine.

coding-bunny commented:
okidoki !
